File: apps/main/src/types/websocket.ts
/**
 * WebSocket Message Envelope Types
 *
 * Defines the structure of messages sent over WebSocket connections.
 * Part of Phase 31.2 - WebSocket Contract & Keep-Alive Tests.
 */

/**
 * Base message envelope for all JSON messages over WebSocket.
 *
 * Plain text messages (ping/pong) bypass this envelope.
 */
export interface WebSocketMessage<T = unknown> {
  /** Message type identifier (e.g., 'connected', 'job:status', 'generation:update') */
  type: string;

  /** Optional message payload */
  payload?: T;

  /** Optional additional data (deprecated - use payload) */
  data?: T;

  /** Optional error information */
  error?: {
    code: string;
    message: string;
  };
}

/**
 * Connection established message
 */
export interface ConnectedMessage extends WebSocketMessage<void> {
  type: 'connected';
  message?: string;
  user_id?: number;
}

/**
 * Generation status update message
 */
export interface GenerationStatusMessage<T = any> extends WebSocketMessage<T> {
  type: 'job:created' | 'job:processing' | 'job:completed' | 'job:failed';
  generation_id?: number;
  status?: string;
  user_id?: number;
  data?: T;
}

/**
 * Error message
 */
export interface ErrorMessage extends WebSocketMessage<void> {
  type: 'error';
  error: {
    code: string;
    message: string;
  };
}

/**
 * Type guard to check if a message is a valid WebSocketMessage
 */
export function isWebSocketMessage(data: unknown): data is WebSocketMessage {
  return (
    typeof data === 'object' &&
    data !== null &&
    'type' in data &&
    typeof (data as any).type === 'string'
  );
}

/**
 * Parse WebSocket message with validation.
 *
 * Returns null for plain text messages (ping/pong keep-alives).
 * Returns parsed message for JSON messages.
 * Throws for invalid JSON or messages without a type field.
 */
export function parseWebSocketMessage(
  raw: string | MessageEvent
): WebSocketMessage | null {
  const data = typeof raw === 'string' ? raw : raw.data;

  // Handle plain text keep-alive messages
  if (data === 'pong' || data === 'ping') {
    return null;
  }

  try {
    const parsed = JSON.parse(data);

    if (!isWebSocketMessage(parsed)) {
      throw new Error('Message missing required "type" field');
    }

    return parsed;
  } catch (err) {
    console.error('[WebSocket] Failed to parse message:', err, 'Raw:', data);
    throw err;
  }
}

File: pixsim7/backend/main/api/v1/websocket.py
"""
WebSocket endpoints for real-time updates

Provides WebSocket connections for generation status updates and other real-time events.
"""
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Depends
from pixsim7.backend.main.api.dependencies import CurrentUser, get_current_user_ws
from pixsim7.backend.main.infrastructure.websocket import connection_manager
from pixsim7.backend.main.infrastructure.websocket.types import ConnectedMessage, is_keep_alive
import logging

logger = logging.getLogger(__name__)

router = APIRouter()


@router.websocket("/ws/generations")
async def websocket_generations(
    websocket: WebSocket,
    token: str = None,
):
    """
    WebSocket endpoint for real-time generation status updates

    Usage:
        Connect to: ws://host/api/v1/ws/generations?token=YOUR_JWT_TOKEN

    Messages sent to client:
        {
            "type": "generation:status",
            "generation_id": 123,
            "status": "processing",
            "user_id": 1,
            "data": {...}  // Full GenerationResponse
        }

    Auth:
        Pass JWT token as query parameter: ?token=YOUR_JWT_TOKEN
    """
    # Extract user from token
    # For now, accept without auth for simplicity (TODO: add proper auth)
    # In production, validate token and get user_id
    user_id = 1  # Placeholder - should come from JWT token validation

    try:
        await connection_manager.connect(websocket, user_id)

        # Send welcome message using typed envelope
        welcome = ConnectedMessage(
            type="connected",
            message="Connected to generation updates",
            user_id=user_id,
        )
        await connection_manager.send_personal_message(
            welcome.model_dump(),
            websocket,
        )

        # Keep connection alive and handle incoming messages
        while True:
            # Wait for messages from client (ping/pong, etc.)
            data = await websocket.receive_text()

            # Handle ping/pong for keep-alive (plain text, not JSON)
            if is_keep_alive(data):
                if data == "ping":
                    await websocket.send_text("pong")
                continue

            # All other messages should be JSON with envelope structure
            # (Add custom message handling here if needed)

    except WebSocketDisconnect:
        connection_manager.disconnect(websocket, user_id)
        logger.info(f"WebSocket disconnected for user {user_id}")
    except Exception as e:
        logger.error(f"WebSocket error for user {user_id}: {e}", exc_info=True)
        connection_manager.disconnect(websocket, user_id)


@router.websocket("/ws/events")
async def websocket_events(
    websocket: WebSocket,
    token: str = None,
):
    """
    WebSocket endpoint for all real-time events

    More generic than /ws/generations - receives all event types.

    Messages sent to client:
        {
            "type": "job:created" | "job:completed" | "asset:created" | ...,
            "data": {...}
        }
    """
    user_id = 1  # Placeholder

    try:
        await connection_manager.connect(websocket, user_id)

        # Send welcome message using typed envelope
        welcome = ConnectedMessage(
            type="connected",
            message="Connected to event stream",
            user_id=user_id,
        )
        await connection_manager.send_personal_message(
            welcome.model_dump(),
            websocket,
        )

        while True:
            data = await websocket.receive_text()

            # Handle ping/pong for keep-alive (plain text, not JSON)
            if is_keep_alive(data):
                if data == "ping":
                    await websocket.send_text("pong")
                continue

            # All other messages should be JSON with envelope structure
            # (Add custom message handling here if needed)

    except WebSocketDisconnect:
        connection_manager.disconnect(websocket, user_id)
    except Exception as e:
        logger.error(f"WebSocket error for user {user_id}: {e}", exc_info=True)
        connection_manager.disconnect(websocket, user_id)


@router.websocket("/ws/jobs")
async def websocket_jobs(
    websocket: WebSocket,
    token: str = None,
):
    """
    WebSocket endpoint for real-time job feed updates

    Provides live updates about:
    - Generation jobs (created, processing, completed, failed)
    - Asset processing jobs
    - Upload jobs
    - Other background tasks

    Usage:
        Connect to: ws://host/api/v1/ws/jobs?token=YOUR_JWT_TOKEN

    Messages sent to client:
        {
            "type": "job:created" | "job:updated" | "job:completed" | "job:failed",
            "job_id": 123,
            "job_type": "generation" | "upload" | "processing",
            "status": "pending" | "processing" | "completed" | "failed",
            "progress": 0.5,  // 0.0 to 1.0
            "data": {...}  // Job-specific data
        }

    Auth:
        Pass JWT token as query parameter: ?token=YOUR_JWT_TOKEN
    """
    user_id = 1  # Placeholder - should come from JWT token validation

    try:
        await connection_manager.connect(websocket, user_id)

        # Send welcome message
        welcome = ConnectedMessage(
            type="connected",
            message="Connected to jobs feed",
            user_id=user_id,
        )
        await connection_manager.send_personal_message(
            welcome.model_dump(),
            websocket,
        )

        # Keep connection alive and handle incoming messages
        while True:
            data = await websocket.receive_text()

            # Handle ping/pong for keep-alive
            if is_keep_alive(data):
                if data == "ping":
                    await websocket.send_text("pong")
                continue

            # Handle other messages if needed

    except WebSocketDisconnect:
        connection_manager.disconnect(websocket, user_id)
        logger.info(f"Jobs WebSocket disconnected for user {user_id}")
    except Exception as e:
        logger.error(f"Jobs WebSocket error for user {user_id}: {e}", exc_info=True)
        connection_manager.disconnect(websocket, user_id)
